src/execute.rs

```diff
@@ -0,0 +1,654 @@
+    let token_provided_remaining: Uint128 =
+        token_provided.amount.checked_sub(refined_token_in.amount)?;
+
+    // Store the remaining amount as a coin
+    let token_provided_remaining_coin: Coin = Coin {
+        denom: token_provided.denom,
+        amount: token_provided_remaining,
+    };
+
+    // Save intermediate state
+    // We will utilize this state after the swap has been executed
+    SWAP_REPLY_STATES.save(
+        deps.storage,
+        SWAP_REPLY_ID,
+        &SwapMsgReplyState {
+            pool_id: pool_id,
+            original_sender: info.sender.clone(),
+            lower_tick: lower_tick,
+            upper_tick: upper_tick,
+            token_min_amount0: token_min_amount0,
+            token_min_amount1: token_min_amount1,
+            token_provided_remaining_coin: token_provided_remaining_coin,
+            token_out_denom: token_out_denom,
+        },
+    )?;
+
+    return Ok(exec_msg);
+}
+
+pub fn calc_amount_0_one_unit_liq(
+    upper_tick: i64,
+    current_tick: i64,
+) -> Result<Decimal256, ContractError> {
+    let p_upper = tick_to_price(upper_tick)?;
+    let p_current = tick_to_price(current_tick)?;
+    let delta_x = Decimal256::one()
+        .checked_mul(p_upper.sqrt().checked_sub(p_current.sqrt())?)?
+        .checked_div(p_upper.sqrt().checked_mul(p_current.sqrt())?)?;
+
+    Ok(delta_x)
+}
+
+pub fn calc_amount_1_one_unit_liq(
+    lower_tick: i64,
+    current_tick: i64,
+) -> Result<Decimal256, ContractError> {
+    let p_lower = tick_to_price(lower_tick)?;
+    let p_current = tick_to_price(current_tick)?;
+    let delta_y = Decimal256::one().checked_mul(p_current.sqrt().checked_sub(p_lower.sqrt())?)?;
+
+    Ok(delta_y)
+}
+
+pub fn calc_asset_ratio_from_ticks(
+    upper_tick: i64,
+    current_tick: i64,
+    lower_tick: i64,
+) -> Result<(Decimal256, Decimal256), ContractError> {
+    if upper_tick < current_tick {
+        return Ok((Decimal256::zero(), Decimal256::one()));
+    }
+    if lower_tick > current_tick {
+        return Ok((Decimal256::one(), Decimal256::zero()));
+    }
+    let delta_x = calc_amount_0_one_unit_liq(upper_tick, current_tick)?;
+    let delta_y = calc_amount_1_one_unit_liq(lower_tick, current_tick)?;
+    let total_delta = delta_x.checked_add(delta_y)?;
+    let asset0_ratio = delta_x.checked_div(total_delta)?;
+    let asset1_ratio = Decimal256::one().checked_sub(asset0_ratio)?;
+
+    Ok((asset0_ratio, asset1_ratio))
+}
+
+fn get_next_sqrt_price_from_amount0_in_round_up(
+    liquidity: Decimal256,
+    sqrt_price_current: Decimal256,
+    token_in: Decimal256,
+) -> Decimal256 {
+    let numerator = liquidity * sqrt_price_current;
+    let denominator = liquidity + (token_in * sqrt_price_current);
+    numerator / denominator
+}
+
+fn get_next_sqrt_price_from_amount1_in_round_down(
+    liquidity: Decimal256,
+    sqrt_price_current: Decimal256,
+    token_in: Decimal256,
+) -> Decimal256 {
+    sqrt_price_current + (token_in / liquidity)
+}
```

Comment:
Note to reviewer: I don't know how to properly round Uint256 to mirror the logic in the osmosis codebase, so I left rounding logic off of these methods

Reply:
Discussed that it's probably fine for v1 version of the contract. v2 can use similar tricks to uniswap math if ever needed